gh-127111: Emscripten Make web example work again

hoodmane opened the pull request:
I moved the web example from `Tools/wasm` into `Tools/wasm/emscripten/web_example`. I also added a new target `build_emscripten` which is `build_wasm` but also builds the web_example. The web_example needs:
1. python.html, copied
2. python.worker.mjs copied
3. python.mjs and python.wasm output from the main linking of the Python interpreter
4. The webserver that sets COOP and COEP
5. python3.14.zip

This last is created by the `wasm_assets.py` script, which required a pretty small set of changes to work fine for us.

The last thing that should be done is the `python.worker.mjs` script should be made independent of the Python version: currently 3.14 is hard coded. [I ran into trouble doing this](https://github.com/emscripten-core/emscripten/issues/22980), so maybe I can leave it to a followup.



<!-- gh-issue-number: gh-127111 -->
* Issue: gh-127111
<!-- /gh-issue-number -->

hoodmane committed:
gh-127111: Make web example work again

I moved the web example from `Tools/wasm` into `Tools/wasm/emscripten/web_example`.
I also added a new target `build_emscripten` which is `build_wasm` but also
builds the web_example. The web_example needs:
1. python.html, copied
2. python.worker.mjs copied
3. python.mjs and python.wasm output from the main linking of the Python interpreter
4. The webserver that sets COOP and COEP
5. python3.14.zip

This last is created by the `wasm_assets.py` script, which required a pretty
small set of changes to work fine for us.

The last thing that should be done is the `python.worker.mjs` script should be
made independent of the Python version: currently 3.14 is hard coded. I ran
into trouble doing this, so maybe I can leave it to a followup.

hoodmane commented on Tools/wasm/emscripten/web_example/server.py:
```diff
@@ -14,13 +14,6 @@
 
 
 class MyHTTPRequestHandler(server.SimpleHTTPRequestHandler):
-    extensions_map = server.SimpleHTTPRequestHandler.extensions_map.copy()
-    extensions_map.update(
-        {
-            ".wasm": "application/wasm",
-        }
-    )
```
This is now the default behavior of `SimpleHTTPRequestHandler`.

hoodmane commented on configure.ac:
```diff
@@ -2321,14 +2325,14 @@ AS_CASE([$ac_sys_system],
 
     dnl Start with 20 MB and allow to grow
-    AS_VAR_APPEND([LDFLAGS_NODIST], [" -sALLOW_MEMORY_GROWTH -sTOTAL_MEMORY=20971520"])
+    AS_VAR_APPEND([LDFLAGS_NODIST], [" -sALLOW_MEMORY_GROWTH -sINITIAL_MEMORY=20971520"])
```
This setting has changed names. It's not really relevant to this PR, I could separate it into a separate NFC commit.

freakboy3742 replied:
I'm fine with including this change as part of this PR. We're still in a "get it working" stage, so we don't gain any real improvement by breaking it out.

hoodmane commented on configure.ac:
```diff
@@ -2321,14 +2325,14 @@ AS_CASE([$ac_sys_system],
     dnl Include file system support
     AS_VAR_APPEND([LDFLAGS_NODIST], [" -sFORCE_FILESYSTEM -lidbfs.js -lnodefs.js -lproxyfs.js -lworkerfs.js"])
-    AS_VAR_APPEND([LDFLAGS_NODIST], [" -sEXPORTED_RUNTIME_METHODS=FS"])
+    AS_VAR_APPEND([LDFLAGS_NODIST], [" -sEXPORTED_RUNTIME_METHODS=FS,callMain"])
```
We use `callMain` in `python.worker.mjs` now so we need to ask for it to be exported.

hoodmane commented on configure.ac:
```diff
@@ -2339,7 +2343,6 @@ AS_CASE([$ac_sys_system],
       AS_VAR_APPEND([LDFLAGS_NODIST], [" -sUSE_PTHREADS"])
       AS_VAR_APPEND([LINKFORSHARED], [" -sPROXY_TO_PTHREAD"])
     ])
-    AS_VAR_APPEND([LDFLAGS_NODIST], [" -sALLOW_MEMORY_GROWTH"])
```
We also set this on line 2328. It's harmless but unnecessary to set it twice.

freakboy3742 commented on Tools/wasm/emscripten/web_example/python.worker.mjs:
```diff
@@ -59,29 +61,42 @@ const stderr = (charCode) => {
         postMessage({type: 'ready', stdinBuffer: stdinBuffer.sab})
+    },
+    async preRun(Module) {
+        // TODO: remove fixed version number
```
Unless it's going to be a major impediment to progress, I think we're going to need a solution for this before we can merge this PR - even though it's an example, it's code that will easily get forgotten along the way.

hoodmane replied:
Well the way that I wanted to do this was blocked by
https://github.com/emscripten-core/emscripten/issues/22980
and I was waiting to see if it got fixed quickly or if I should find a workaround. Emscripten 3.1.73 was released with a fix so I think we can bump the minimum suggested Emscripten version to 3.1.73 and use the straightforward approach.

hoodmane replied:
I suppose one simple fix would be to template the version into `python.sh` and then pass it in as an argument or environment variable. But I was thinking to get it from `Py_Version`.

hoodmane replied:
Okay I fixed this in 7c61ed16707e6128b352575d57b6d267538ef3bc.

freakboy3742 reviewed (changes requested):
The general shape of this makes sense. 

From a general process perspective, this needs a news entry to describe that that web example has been updated, and documentation in the WASM README on how to run the example.

However, assuming the usage instructions are `python Tools/wasm/emscripten build` and open `web_example/python.html` - I get 2 errors.
 
First, during the build process, the macOS version of `realpath` apparently doesn't honor the `-s` flag, so the usage on L223 of the `Tools/emscripten/__main__.py` raises an error and `python Tools/wasm/emscripten build` doesn't complete.

~Second, I wasn't able to see the example working in practice. When I open `file:///.../web_example/python.html` in Chrome, I get the following error:~
```
python.html:50 Uncaught (in promise) SecurityError: Failed to construct 'Worker': Script at 'file:///Users/rkm/projects/python/cpython/cross-build/wasm32-emscripten/web_example/python.worker.mjs' cannot be accessed from origin 'null'.
    at WorkerManager.initialiseWorker (python.html:50:27)
    at new WorkerManager (python.html:45:14)
    at window.onload (python.html:350:33)
```
~If I use `python -m http.server` to serve the `web_example` folder, and open python.html, I get:~
```
Uncaught ReferenceError: SharedArrayBuffer is not defined
    at new StdinBuffer (python.worker.mjs:5:24)
    at python.worker.mjs:62:21
```

EDIT: Strike that - I see from @JeffersGlass comment that `python web_example/server.py` is the expected operation; that works for me. That only leaves the issue with `realpath` failing during  the build (that's a problem that I didn't pick up with my review of #126903, but we can fix it here)

hoodmane committed:
Add workaround for defective mac coreutils

hoodmane committed:
Update readme

hoodmane committed:
Add news entry

hoodmane committed:
Double backticks

freakboy3742 commented on Tools/wasm/README.md:
```diff
@@ -70,6 +70,78 @@ instance, to do a debug build, you can use:
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements
+
+If you serve the web example with ``python server.py`` and then visit
+``localhost:8000`` in a browser it should work.
```
```suggestion
To run the web example, ``cd`` into the ``web_example`` directory, then run 
``python server.py``. This will start a web server; you can then visit 
``http://localhost:8000/python.html`` in a browser to see a simple REPL example.
```

Also - I think this fits better above before "The web example uses SharedArrayBuffer...". The most important detail is the "prove it works" example usage; details about the specifics of the example are interesting once they've proved it works, and want to adapt it for their own usage.

hoodmane replied:
Indeed, makes sense.

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
+                REALPATH=abs_path
+            fi
+            """
+            f"""\
```
The `\` here leads to some odd indentation of the first comment line.
```suggestion
            f"""
```

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
             #!/bin/sh
 
+            # Macs come with a defective fork of coreutils so feature detect and
+            # work around it.
```
This is unnecessarily pejorative. macOS ships with the [FreeBSD version of coreutils](https://man.freebsd.org/cgi/man.cgi?query=realpath&sektion=1). While I agree the discrepancies between GNU and FreeBSD are often annoying, and in this instance the GNU implementation has useful features, that doesn't make the FreeBSD version a "defective fork".

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
+            if which grealpath > /dev/null; then
+                # It has brew installed gnu core utils, use that
+                REALPATH="grealpath -s"
+            elif which readlink > /dev/null && realpath --version | grep GNU > /dev/null; then
```
`--version` isn't a valid option for FreeBSD's realpath, so this generates output on stderr.
```suggestion
            elif which realpath > /dev/null && realpath --version 2&>1 | grep GNU > /dev/null; then
```

freakboy3742 commented on Tools/wasm/README.md:
```diff
@@ -70,6 +70,78 @@ instance, to do a debug build, you can use:
+### Running from node
+
+If you want to run the normal Python cli, you can use `python.sh`. It takes the
+same options as the normal Python cli entrypoint, though the REPL does not
```
```suggestion
If you want to run the normal Python CLI, you can use `python.sh`. It takes the
same options as the normal Python CLI entrypoint, though the REPL does not
```

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
+            else
+                # Shim for macs without GNU coreutils
+                abs_path () {
+                    echo "$(cd $(dirname "$1");pwd)/$(basename "$2")"
```
Should this be:
```suggestion
                    echo "$(cd $(dirname "$1");pwd)/$(basename "$1")"
```
?

However, even then, if I try to run a `hello.py` (that contains `print("hello, world")`) that is in the root directory of the project, I get:
```
(venv) rkm@eunectes wasm32-emscripten % ./python.sh hello.py
realpath: illegal option -- -
usage: realpath [-q] [path ...]
/Users/rkm/projects/python/cpython/cross-build/wasm32-emscripten/python.sh: can't open file '//hello.py': [Errno 44] No such file or directory
```
Am I misunderstanding the intended usage?

hoodmane replied:
Yeah... This is currently the expected output because of other bugs ("limitations") in `node_entry.mjs`. I wasn't planning to fix them in this PR because it's not relevant to the description.

The problems are:
1. Working directory not mounted into emscripten file system so it can't see `some_file.py`
2. Working directory not set correctly in emscripten so even if it were mounted it would look in the wrong place
3. Environment variables not propagated so in particular `HOME` is weird

Before then it won't work all that well like the python cli but with these changes it works pretty okay. I'm still also waiting on https://github.com/emscripten-core/emscripten/pull/22925 to get merged so that named pipes won't cause chaos.

Maybe I should make a separate PR and move the bsd realpath fixes and these three fixes there.

hoodmane replied:
https://github.com/python/cpython/pull/127506

freakboy3742 reviewed (changes requested):
A couple of documentation tweaks; I'm still having issues with realpath as well. Not sure if I'm doing something wrong, though.

hoodmane committed:
Adjust readme

hoodmane committed:
Update shell scripts

Apply shell check fixes
don't disparage BSD coreutils
fix indentation
realpath not readlink

hoodmane committed:
Capitalize CLI

hoodmane committed:
Don't hard code version

hoodmane committed:
Bump recommended Emscripten version

hoodmane committed:
Revert changes to emscripten/__main__.py

freakboy3742 reviewed (approved):
This obviously needs the changes from #127506 to be fully usable on macOS; but I'm happy with this as a set of changes in itself.
